File: ios/XiangshangYouth/Core/Services/FrontendTelemetry.swift
import Foundation
import OSLog

enum FrontendEvent: String {
    case growthReportOpened = "growth_report_opened"
    case growthReportPeriodChanged = "growth_report_period_changed"
    case adaptivePlanOpenedCourses = "adaptive_plan_opened_courses"
}

enum FrontendTelemetry {
    private static let sink = ProductEventSink()

    static func configure(enabled: Bool) {
        Task { await sink.configure(enabled: enabled) }
    }
    static func record(_ event: FrontendEvent, value: String? = nil) {
        Task { await sink.record(event, value: value) }
    }
}

private actor ProductEventSink {
    private let logger = Logger(subsystem: "com.xiangshang.youth", category: "product-events")
    private let clientSessionID = UUID()
    private var enabled = false

    func configure(enabled: Bool) { self.enabled = enabled }

    func record(_ event: FrontendEvent, value: String?) async {
        guard enabled else { return }
        let allowedValue = ["本周", "本月"].contains(value) ? value : nil
        logger.info("event=\(event.rawValue, privacy: .public) value=\(allowedValue ?? "none", privacy: .public)")
        let payload = ProductEventInput(
            eventID: UUID(), eventName: event.rawValue, coarseValue: allowedValue,
            platform: "ios", appVersion: Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "1.0",
            clientSessionID: clientSessionID, occurredAt: ISO8601DateFormatter().string(from: Date())
        )
        // Analytics must never block, retry, or alter a family/teacher flow.
        try? await ProductEventApi.submit([payload])
    }
}

File: android/app/src/main/java/com/xiangshang/youth/core/util/FrontendTelemetry.kt
package com.xiangshang.youth.core.util

import android.util.Log
import com.xiangshang.youth.BuildConfig
import com.xiangshang.youth.core.service.ApiClient
import com.xiangshang.youth.core.service.ProductEventApi
import com.xiangshang.youth.core.service.ProductEventBatch
import com.xiangshang.youth.core.service.ProductEventInput
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import java.time.Instant
import java.util.UUID

enum class FrontendEvent(val wireName: String) {
    GrowthReportOpened("growth_report_opened"),
    GrowthReportPeriodChanged("growth_report_period_changed"),
    AdaptivePlanOpenedCourses("adaptive_plan_opened_courses")
}

/**
 * Privacy-safe event boundary. Values must be coarse UI states and must never
 * include student ids, names, phone numbers or health measurements.
 */
object FrontendTelemetry {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val clientSessionId = UUID.randomUUID().toString()
    @Volatile private var enabled = false

    fun configure(enabled: Boolean) { this.enabled = enabled }

    fun record(event: FrontendEvent, value: String? = null) {
        if (!enabled) return
        val coarse = value?.takeIf { it == "本周" || it == "本月" }
        Log.i("XiangshangEvents", "event=${event.wireName} value=${coarse ?: "none"}")
        val item = ProductEventInput(
            UUID.randomUUID().toString(), event.wireName, coarse, "android",
            BuildConfig.VERSION_NAME, clientSessionId, Instant.now().toString()
        )
        // Product analytics is non-critical: it cannot block or mutate a
        // family, teacher, report, assessment or training workflow.
        scope.launch {
            runCatching { ApiClient.retrofit.create(ProductEventApi::class.java).submit(ProductEventBatch(listOf(item))) }
        }
    }
}
